### PR TITLE
Issue2 fixed

### LINKED ISSUE #3: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [x] What is a valid image URL?

__Reference material__:
- https://github.com/plataformatec/simple_form

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by theovelikov**
Closes #3 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -206,6 +209,9 @@ GEM
     uglifier (4.1.20)
       execjs (>= 0.3.0, < 3)
     unicode-display_width (1.6.0)
+    validate_url (1.0.11)
+      activemodel (>= 3.0.0)
+      public_suffix
```
Comment (YJ83Lee) [inline Gemfile.lock:218]: This is minor.

I believe this part goes with adding `validate_url` to `Gemfile` in the next commit.

File: Gemfile.lock [hunk 2]
```diff
@@ -235,11 +241,13 @@ DEPENDENCIES
   rails (~> 5.2.0)
   sass-rails (~> 5.0, >= 5.0.6)
   selenium-webdriver
+  simple_form
   spring
   spring-watcher-listen (~> 2.0.0)
   sqlite3
   tzinfo-data
   uglifier (>= 1.3.0)
+  validate_url
```
Comment (YJ83Lee) [inline Gemfile.lock:255]: Same as the comment above.

File: config/routes.rb [hunk 3]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root 'application#home'
+  get 'application/home'
+  resources :images
```
Comment (YJ83Lee) [inline config/routes.rb]: I see that routes to controller actions not implemented are restricted here: https://github.com/AppFolioOnboarding/image-sharer-theovelikov/pull/21/commits/c0156e6cf2da1625435459aed729d08595de94b0#diff-21497849d8f00507c9c8dcaf6288b136R5, which is good.

File: config/spring.rb [hunk 4]
```diff
@@ -4,3 +4,11 @@
...
   tmp/caching-dev.txt
 ].each { |path| Spring.watch(path) }
+Spring.after_fork do
+  if ENV['DEBUGGER_STORED_RUBYLIB']
+    ENV['DEBUGGER_STORED_RUBYLIB'].split(File::PATH_SEPARATOR).each do |path|
+      next unless path =~ /ruby-debug-ide/
+      load path + '/ruby-debug-ide/multiprocess/starter.rb'
+    end
+  end
+end
```
Comment (YJ83Lee) [inline config/spring.rb]: Seems like these lines are added after we tried to use RubyMine debugger.

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    @image = Image.new
+  end
+
+  def create
+    @image = Image.new(image_params)
+
+    @image.save
+    redirect_to @image
+
+  end
```
Comment (YJ83Lee) [inline app/controllers/images_controller.rb]: Controller tests are in this commit: https://github.com/AppFolioOnboarding/image-sharer-theovelikov/pull/21/commits/c0156e6cf2da1625435459aed729d08595de94b0#diff-8807b315bee3099305dbdee3d4f0a4f7R1-R26.

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
+require 'test_helper'
+
+class ImagesControllerTest < ActionDispatch::IntegrationTest
+  test "show_test" do
+    image = Image.create!(url:"https://www.humanesociety.org/sites/default/files/styles/1240x698/public/2018/08/kitten-440379.jpg?h=c8d00152&itok=1fdekAh2")
+    get image_path(image.id)
+    assert_response :success
+  end
```
Comment (YJ83Lee) [inline test/controllers/images_controller_test.rb]: This is minor.

What do you think about the case where the id in the URL is invalid (i.e., when there is no row of `images` in the DB with the given `id`).

File: config/initializers/simple_form.rb [hunk 7]
```diff
@@ -14,7 +15,7 @@
   # stack. The options given below are used to wrap the
   # whole input.
   config.wrappers :default, class: :input,
-    hint_class: :field_with_hint, error_class: :field_with_errors, valid_class: :field_without_errors do |b|
+                            hint_class: :field_with_hint, error_class: :field_with_errors, valid_class: :field_without_errors do |b| # rubocop:disable Metrics/LineLength
```
Comment (YJ83Lee) [inline config/initializers/simple_form.rb:18]: We probably do not need to fix style for the file provided by simple form.

File: Gemfile [hunk 8]
```diff
@@ -47,6 +47,7 @@ end
 group :test do
   gem 'capybara'
   gem 'selenium-webdriver'
+  gem 'rails-controller-testing'
```
Comment (YJ83Lee) [inline Gemfile:49]: I believe we added this gem to use `assert_template`?

File: test/controllers/images_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    assert_equal Image.last.url, "https://www.humanesociety.org/sites/default/files/styles/1240x698/public/2018/08/kitten-440379.jpg?h=c8d00152&itok=1fdekAh2"
+  end
+
+  test "create_test_invalid" do
+    post images_path, params: {image:{url:" "}}
+    assert_response :success
+    assert_template :new
+  end
+
+  test "new_test" do
```
Comment (YJ83Lee) [inline test/controllers/images_controller_test.rb]: This is minor.

In controller test sometimes you might want to use `assert_select` to make sure the rendered template does contains elements you want to have, e.g., input field for url.

File: config/routes.rb [hunk 10]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root 'application#home'
+  get 'application/home'
```
Comment (YJ83Lee) [inline config/routes.rb:4]: I am not sure if we need to add this route?

File: Gemfile (hunk 8 above)
Comment (theovelikov) [reply Gemfile:49]: Yes, after talking to Natalie she asked me to use a different assert statement

**Pull request merged**